thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp

```diff
@@ -1332,51 +1348,87 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
     {
         typedef CKeyLookupLocalBase PARENT;
 
-        std::vector<IKeyManager *> keyManagers;
+        //Need ptr to std::vector as std::atomic's are not constructable(doesn't have copy constructor)
+        std::unique_ptr<std::vector<std::atomic<IKeyManager *>>> keyManagers;
+        CriticalSection cs;
+
     public:
         CKeyLookupLocalHandler(CKeyedJoinSlave &_activity) : CKeyLookupLocalBase(_activity)
         {
         }
         ~CKeyLookupLocalHandler()
         {
-            for (auto &km : keyManagers)
+            for (auto &km : *keyManagers)
             {
                 if (km)
-                    km->Release();
+                    (*km).Release();
             }
         }
+        virtual void init() override
+        {
+            PARENT::init();
+            keyManagers.reset(new std::vector<std::atomic<IKeyManager *>>(parts.size()));
+            for (auto & k: *keyManagers)
+                k = nullptr;
+        }
         virtual void addPartNum(unsigned partNum) override
         {
             PARENT::addPartNum(partNum);
-            keyManagers.push_back(nullptr);
             translators.push_back(nullptr);
         }
         virtual void process(CThorExpandingRowArray &processing, unsigned selected) override
         {
             unsigned partCopy = parts[selected];
             unsigned partNo = partCopy & partMask;
             unsigned copy = partCopy >> partBits;
-            IKeyManager *&keyManager = keyManagers[selected];
+            IKeyManager *keyManager = (*keyManagers)[selected];
             if (!keyManager) // delayed until actually needed
             {
-                keyManager = activity.createPartKeyManager(partNo, copy);
-                // NB: potentially translation per part could be different if dealing with superkeys
-                setupTranslation(partNo, selected, *keyManager);
+                CriticalBlock b(cs);
```

Comment:
Was there an existing bug where process could be called on multiple threads?  If not, is the criticalBlock needed?

Reply:
There is only 1 process() thread for this object and it's the only thread that updates the keyManagers array, so shouldn't need CS/double check.

I think as before, you should take the vector item by reference and assign, except now it's an atomic. e.g.:
```
            std::atomic<IKeyManager *> &keyManager = (*keyManagers)[selected];
            if (!keyManager) // delayed until actually needed
            {
                keyManager = activity.createPartKeyManager(partNo, copy);
```